src/eckit/types/Time.cc

```diff
@@ -8,34 +8,64 @@
  * does it submit to any jurisdiction.
  */
 
-#include <regex>
+#include <cctype>
 #include <cmath>
+#include <cstddef>
+#include <regex>
+#include <string>
+#include <string_view>
 
-#include "eckit/eckit.h"
-
+#include "eckit/exception/Exceptions.h"
 #include "eckit/persist/DumpLoad.h"
 #include "eckit/types/Time.h"
 #include "eckit/utils/Hash.h"
-#include "eckit/utils/Tokenizer.h"
 
 namespace {
-    static thread_local std::regex digits_time_("^-?[0-9]+$");
-    static thread_local std::regex float_hours_("^-?[0-9]*\\.[0-9]+$");
-    static thread_local std::regex hhmmss_("^([0-9]+):([0-5]?[0-9])(:[0-5]?[0-9])?$");
-    static thread_local std::regex ddhhmmss_("^-?([0-9]+[dD])?([0-9]+[hH])?([0-9]+[mM])?([0-9]+[sS])?$");
-}
-
-namespace eckit {
 
 //----------------------------------------------------------------------------------------------------------------------
 
-inline void printTime(std::ostream& s, long n) {
+const std::regex hhmmss_("^([0-9]+):([0-5]?[0-9])(:[0-5]?[0-9])?$");
+const std::regex ddhhmmss_("^-?([0-9]+[dD])?([0-9]+[hH])?([0-9]+[mM])?([0-9]+[sS])?$");
+
+// DIGITS: "^-?[0-9]+$"
+// FLOAT:  "^-?[0-9]*\\.[0-9]+$"
+enum class TimeFormat { UNKOWN, OTHER, DIGITS, DECIMAL };
+
+TimeFormat checkTimeFormat(const std::string_view time) {
+    bool hasDigit   = false;
+    bool hasDecimal = false;
+
+    const std::size_t start = (time[0] == '-') ? 1 : 0;
+
+    for (auto i = start; i < time.length(); i++) {
+        if (time[i] == '.') {
+            if (hasDecimal || i == time.length() - 1) { return TimeFormat::UNKOWN; }
+            hasDecimal = true;
+        } else if (isdigit(time[i]) == 0) {
+            return TimeFormat::OTHER;
+        } else {
+            hasDigit = true;
+        }
+    }
```

Comment:
I think this can be done with a single pass over the input string and then return the typeid of the found variant (pretty similar to what you are doing already) plus up to 5 tokens, 1 token for the sign and 4 for positive integers (day, hour, minute, second).

Subsequent code can then validate on this result, e.g. was the extended flag passed.

I am thinking along the lines of:
```cpp
struct tokenized_time {
    FormatType type;
    std::string_view sign;
    std::string_view integers[4];
};
```